page-tracking/src/page_tracker.rs

```diff
@@ -606,24 +606,42 @@ impl HypPageAlloc {
         dirty_pages.clean()
     }
 
-    /// Same as above, but the returned pages are assigned for internal use.
-    pub fn take_pages_for_host_state_with_alignment(
+    // Same as `take_pages`, but returned page are state pages for either hypervisor or host.
+    //
+    // Safety:
+    // `owner` must be either `PageOwnerId::host()` or `PageOwnerId::hypervisor()`.
+    unsafe fn take_state_pages(
         &mut self,
+        owner: PageOwnerId,
         count: usize,
         align: u64,
     ) -> SequentialPages<InternalClean> {
         let assignable_pages = self.take_pages(count, align);
         SequentialPages::from_pages(assignable_pages.into_iter().map(|p| {
             let page_info = self.pages.get_mut(p.addr()).unwrap();
-            page_info
-                .assign(PageOwnerId::host(), PageState::VmState)
-                .unwrap();
+            page_info.assign(owner, PageState::State).unwrap();
```

Comment:
The pages returned by `take_pages()` are already hypervisor-owned, so in the `take_pages_for_hyp_state()` case you end up pushing the hypervisor as the owner again. What I think you want instead is to make `take_pages()` more generic w.r.t. the state it sets for the newly-claimed pages.

Reply:
Okay, makes sense.